python/tvm/topi/cuda/argwhere.py

```diff
@@ -0,0 +1,621 @@
+    a0 = condition.shape[0]
+
+    condition = ib.buffer_ptr(condition)
+    out = ib.buffer_ptr(out)
+
+    valid_index = ib.allocate("int32", (1,), name="valid_index", scope="global")
+    tmp = ib.allocate("int32", (1,), name="tmp", scope="local")
+    one_count = tvm.tir.const(1, dtype="int32")
+
+    max_threads = int(tvm.target.Target.current(allow_none=False).max_num_threads)
+    nthread_tx = max_threads
+    # Limit threads to a single block to make sure atomic_add works normally.
+    tx = te.thread_axis("threadIdx.x")
+    ib.scope_attr(tx, "thread_extent", nthread_tx)
+    len_inner_for = a0 // nthread_tx + 1
+    valid_index[0] = 0
+
+    with ib.for_range(0, len_inner_for, name="i") as i:
+        idx = tx * len_inner_for + i
+        with ib.if_scope(idx < a0):
+            with ib.if_scope(condition[idx] != 0):
+                tmp[0] = atomic_add(
+                    tvm.tir.call_intrin("handle", "tir.address_of", valid_index[0]),
+                    one_count,
+                )
+                out[tmp[0]] = idx
+
+    return ib.get()
+
+
+def argwhere_1d(output_shape, condition):
+    """Compute for argwhere 1D
+
+    Parameters
+    ----------
+    condition : list of int or tvm.tir.Any
+        The output shape
+
+    out : tvm.te.Tensor
+        Tensor with boolean values.
+
+    Returns
+    -------
+    stmt : Stmt
+        The result IR statement.
+    """
+    condition_buf = tvm.tir.decl_buffer(
+        condition.shape, condition.dtype, "data_buf", data_alignment=8
+    )
+    out_buf = tvm.tir.decl_buffer(output_shape, "int32", "out_buf", data_alignment=8)
+
+    out = te.extern(
+        [output_shape],
+        [condition],
+        lambda ins, outs: argwhere_1d_ir(ins[0], outs[0]),
+        dtype=["int32"],
+        in_buffers=[condition_buf],
+        out_buffers=[out_buf],
+        name="argwhere_1d",
+        tag="argwhere1d_gpu",
+    )
+
+    if out.shape[0] <= 1:
+        return out
+
+    sorted_out = _get_sort_func()(
+        out, k=0, axis=0, ret_type="values", is_ascend="True", dtype="int32"
+    )
+
+    return sorted_out
+
+
+def argwhere_2d_ir(condition, out):
+    """Low level IR for argwhere 2D
+
+    Parameters
+    ----------
+    condition : Buffer
+        The condition buffer.
+
+    out : Buffer
+        The output buffer.
+
+    Returns
+    -------
+    stmt : Stmt
+        The result IR statement.
+    """
+    ib = tvm.tir.ir_builder.create()
+    a0 = condition.shape[0]
+    a1 = condition.shape[1]
+
+    condition = ib.buffer_ptr(condition)
+    out = ib.buffer_ptr(out)
+
+    valid_index = ib.allocate("int32", (1,), name="valid_index", scope="local")
+    tmp = ib.allocate("int32", (1,), name="tmp", scope="local")
+    one_count = tvm.tir.const(1, dtype="int32")
+
+    max_threads = int(tvm.target.Target.current(allow_none=False).max_num_threads)
+    nthread_tx = max_threads
+
+    # Limit threads to a single block to make sure atomic_add works normally.
```

Comment:
Cuda does have a kernel level atomic add (https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#atomic-functions). It is just slower or do we not have access to it from TIR?

Reply:
We use ```atomicAdd```. However, if number of blocks is larger than a threshold, it will return incorrect result.

Reply:
I'm surprised. I'd expect atomicAdd to work with any number of blocks. Could you maybe expand this comment with why and when atomicAdd fails?

Reply:
The observation is that if input data size is large( > 300 * 300 for example), previous we don't limit the number of blocks and the output of IR routine would be incorrect. I didn't dig deeper into it at this time.